File: java/SelectionSort.java
public class SelectionSort {
    void selection(int[] a) /* function to sort an array with selection sort */ {
        int i, j, small;
        int n = a.length;
        for (i = 0; i < n - 1; i++) {
            small = i; //minimum element in unsorted array

            for (j = i + 1; j < n; j++)
                if (a[j] < a[small])
                    small = j;
            // Swap the minimum element with the first element
            int temp = a[small];
            a[small] = a[i];
            a[i] = temp;
        }

    }

    void printArr(int[] a) /* function to print the array */ {
        int i;
        int n = a.length;
        for (i = 0; i < n; i++)
            System.out.print(a[i] + " ");
    }

    public static void main(String[] args) {
        int[] a = {91, 49, 4, 19, 10, 21};
        SelectionSort i1 = new SelectionSort();
        System.out.println("\nBefore sorting array elements are - ");
        i1.printArr(a);
        i1.selection(a);
        System.out.println("\nAfter sorting array elements are - ");
        i1.printArr(a);
        System.out.println();
    }
}

File: js/SelectionSort.js
"use strict";
// Selection sort with O(n^2) time complexity

function Selection_Sort(arr, compare_Function) {
  function compare(a, b) {
    return a - b;
  }
  let min = 0;
  let index = 0;
  let temp = 0;

  compare_Function = compare_Function || compare;

  for (let i = 0; i < arr.length; i += 1) {
    index = i;
    min = arr[i];

    for (let j = i + 1; j < arr.length; j += 1) {
      if (compare_Function(min, arr[j]) > 0) {
        min = arr[j];
        index = j;
      }
    }

    temp = arr[i];
    arr[i] = min;
    arr[index] = temp;
  }

  return arr;
}

console.log(
  Selection_Sort([3, 0, 2, 5, -1, 4, 1], function (a, b) {
    return a - b;
  })
);
console.log(
  Selection_Sort([3, 0, 2, 5, -1, 4, 1], function (a, b) {
    return b - a;
  })
);
